src/app/codegen-interaction-model/CodegenDataModel.cpp

```diff
@@ -0,0 +1,411 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/codegen-interaction-model/CodegenDataModel.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/endpoint-config-api.h>
+#include <lib/core/DataModelTypes.h>
+#include <optional>
+
+namespace chip {
+namespace app {
+namespace {
+
+/// Checks if the specified ember cluster mask corresponds to a valid
+/// server cluster.
+bool IsServerMask(EmberAfClusterMask mask)
+{
+    return (mask == 0) || ((mask & CLUSTER_MASK_SERVER) != 0);
+}
+
+/// Load the cluster information into the specified destination
+void LoadClusterInfo(const ConcreteClusterPath & path, const EmberAfCluster & cluster, InteractionModel::ClusterInfo * info)
+{
+    chip::DataVersion * versionPtr = emberAfDataVersionStorage(path);
+    if (versionPtr != nullptr)
+    {
+        info->dataVersion = *versionPtr;
+    }
+    else
+    {
+        ChipLogError(AppServer, "Failed to get data version for %d/" ChipLogFormatMEI, static_cast<int>(path.mEndpointId),
+                     ChipLogValueMEI(cluster.clusterId));
+        info->dataVersion = 0;
+    }
+
+    // TODO: set entry flags:
+    //   info->flags.Set(ClusterQualityFlags::kDiagnosticsData)
+}
+
+/// Converts a EmberAfCluster into a ClusterEntry
+InteractionModel::ClusterEntry ClusterEntryFrom(EndpointId endpointId, const EmberAfCluster & cluster)
+{
+    InteractionModel::ClusterEntry entry;
+
+    entry.path = ConcreteClusterPath(endpointId, cluster.clusterId);
+    LoadClusterInfo(entry.path, cluster, &entry.info);
+
+    return entry;
+}
+
+/// Finds the first server cluster entry for the given endpoint data starting at [start_index]
+///
+/// Returns an invalid entry if no more server clusters are found
+InteractionModel::ClusterEntry FirstServerClusterEntry(EndpointId endpointId, const EmberAfEndpointType * endpoint,
+                                                       unsigned start_index, unsigned & found_index)
+{
+    for (unsigned cluster_idx = start_index; cluster_idx < endpoint->clusterCount; cluster_idx++)
+    {
+        const EmberAfCluster & cluster = endpoint->cluster[cluster_idx];
+        if (!IsServerMask(cluster.mask & CLUSTER_MASK_SERVER))
+        {
+            continue;
+        }
+
+        found_index = cluster_idx;
```

Comment:
All the callers pass mClusterIterationHint for found_index.  This is fine, but we could also just set mClusterIterationHint here and not have the extra arg.

Reply:
This would make this a member method, right now it is not. I think the overhead is probably not too bad for now.